hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FairScheduler.java

```diff
@@ -987,12 +1010,18 @@ void attemptScheduling(FSSchedulerNode node) {
       }
 
       // Assign new containers...
-      // 1. Check for reserved applications
-      // 2. Schedule if there are no reservations
-
-      boolean validReservation = false;
+      // 1. Ensure containers are assigned to the apps that preempted
+      // 2. Check for reserved applications
+      // 3. Schedule if there are no reservations
+
+      // Apps may wait for preempted containers
+      // We have to satisfy these first to avoid cases, when we preempt
+      // a container for A from B and C gets the preempted containers,
+      // when C does not qualify for preemption itself.
+      boolean validReservation;
+      validReservation = assignPreemptedContainers(node);
       FSAppAttempt reservedAppSchedulable = node.getReservedAppSchedulable();
-      if (reservedAppSchedulable != null) {
+      if (!validReservation && reservedAppSchedulable != null) {
```

Comment:
In lieu of future changes around assigning as many containers as directed by assign-multiple settings, can we not gate further allocations on the result of assignPreemptedContainers.